app/src/main/java/com/scouts/kitchenplaner/model/usecases/DisplayRecipeForCooking.kt

```diff
@@ -0,0 +1,111 @@
+
+import com.scouts.kitchenplaner.datalayer.repositories.AllergenRepository
+import com.scouts.kitchenplaner.datalayer.repositories.RecipeRepository
+import com.scouts.kitchenplaner.model.entities.DietarySpeciality
+import com.scouts.kitchenplaner.model.entities.DietaryTypes
+import com.scouts.kitchenplaner.model.entities.MealSlot
+import com.scouts.kitchenplaner.model.entities.Project
+import com.scouts.kitchenplaner.model.entities.RecipeAlternative
+import com.scouts.kitchenplaner.model.entities.RecipeForCooking
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.combine
+
+/**
+ * Usecase for displaying a recipe for cooking
+ *
+ * @param recipeRepository Repository class to retrieve information about recipes
+ * @param allergenRepository Repository class to retrieve information about allergen persons
+ */
+class DisplayRecipeForCooking(
+    private val recipeRepository: RecipeRepository,
+    private val allergenRepository: AllergenRepository
+) {
+    /**
+     * Retrieve information about a recipe for cooking
+     *
+     * @param project The project in which the recipe is cooked
+     * @param mealSlot The meal slot for which the recipe is cooked
+     * @param recipeID The ID of the recipe being cooked
+     *
+     * @return A flow containing information on how to cook the recipe
+     */
+    fun showRecipeForCooking(
+        project: Project,
+        mealSlot: MealSlot,
+        recipeID: Long
+    ): Flow<RecipeForCooking> {
+        val allergenPersonsFlow = allergenRepository.getAllergenPersonsByProjectID(project.id)
+        val allergensFlow = recipeRepository.getAllergensForRecipe(recipeID)
+        val alternativeRecipes =
+            project.mealPlan[mealSlot].first?.second?.map { recipeRepository.getRecipeById(it.id) }
+                ?: listOf()
+
+        val combinedAlternatives = combine(alternativeRecipes) { it.toList() }
+
+        val allergensForAlternatives = combine(
+            combinedAlternatives,
+            allergensFlow,
+            allergenPersonsFlow
+        ) { alternatives, allergens, allergenPersons ->
+            val allergenMap = mutableMapOf<Long, List<DietarySpeciality>>()
+            alternatives.forEach {
+                val relevantTraces = it.traces
+                    .filter { trace ->
+                        allergens.any { speciality ->
+                            speciality.allergen == trace && speciality.type == DietaryTypes.ALLERGEN
+                        }
+                    }
+                    .map { name -> Pair(name, DietaryTypes.TRACE) }
+
+                val relevantFrees = it.freeOfAllergen
+                    .filter { free ->
+                        allergens.any { speciality ->
+                            speciality.allergen == free && speciality.type != DietaryTypes.FREE_OF
+                        }
+                    }
+                    .map { name -> Pair(name, DietaryTypes.FREE_OF) }
```

Comment:
What if there the alternative improves on some ingredients on the alternative but is worse on another ingredient that the recipe to display is free of?

Line 86: all

Reply:
I dont understand the question

Reply:
And no, line 86 should not be all, an allergen is not only relevant if all allergens of the main recipe have the same name as it.